### PR TITLE
Integrate lidar

### PR DESCRIPTION
**Pull request by xandercai**
DESCRIPTION OF PR:

## Developer Checklist
- [x ] Make code change
- [ ] Update tests
  - [ ] Update / create new tests
  - [ ] Ensure these tests have the expected behaviour
  - [ ] Test locally and ensure tests are passing
- [ ] Update documentation
  - [ ] Readme
  - [ ] Docstrings
  - [ ] Comments
  - [ ] Wiki

## Reviewer Checklist
- [ ] Check new code for code smells
- [ ] Check new tests
  - [ ] Ensure adequate coverage
  - [ ] Check for code smells within tests
- [ ] Check if documentation needs updating
  - [ ] Readme
  - [ ] Docstrings
  - [ ] Comments
  - [ ] Wiki

### CODE REVIEW

File: src/dynamic_boundary_conditions/rec1_osm_match.py [hunk 1]
```diff
@@ -15,7 +15,7 @@
 from shapely.geometry import Point
 from sqlalchemy.engine import Engine
 
-from src.lidar import dem_metadata_in_db
+from newzealidar.utils import get_dem_band_and_resolution_by_geometry
```
Comment (LukeParky) [inline src/dynamic_boundary_conditions/rec1_osm_match.py:18]:
```suggestion
import sys
sys.path.insert(0, r'../NewZeaLiDAR')
from newzealidar.utils import get_dem_band_and_resolution_by_geometry
```
Needs this to be able to run

File: src/flood_model/bg_flood_model.py [hunk 2]
```diff
@@ -398,10 +403,16 @@ def run_bg_flood_model(
...
+    # hydro_dem_path = tables.get_data_by_id(engine, tables.DEM, index, geom_col='')['hydro_dem_path'].values[0]
+    df = tables.get_data_by_id(engine, tables.DEM, index, geom_col='')
+    hydro_dem_path = pathlib.Path(df['hydro_dem_path'].values[0])
     # Use the resolution of the Hydro DEM if it is not provided
     if resolution is None:
-        _, resolution = dem_metadata_in_db.get_hydro_dem_data_and_resolution(engine, catchment_area)
+        instructions_file = pathlib.Path(utils.get_env_variable("INSTRUCTIONS_FILE"))
+        with open(instructions_file, 'r') as f:
+            instructions = json.loads(f.read())
+            resolution = instructions['instructions']['output']['grid_params']['resolution']
```
Comment (LukeParky) [inline src/flood_model/bg_flood_model.py:415]: As discussed in meeting yesterday, this resolution should be gathered from either the hydrological DEM or if you think it is a waste of resources then it should be read from a new column in the database

File: src/run_all.py [hunk 3]
```diff
@@ -25,8 +28,8 @@ def main(selected_polygon_gdf: gpd.GeoDataFrame, modules_with_log_levels: Dict[M
     # Define a dictionary mapping each module to its log level
     module_to_log_level = {
         run: LogLevel.DEBUG,
-        lidar_metadata_in_db: LogLevel.DEBUG,
-        dem_metadata_in_db: LogLevel.DEBUG,
+        datasets: LogLevel.DEBUG,  # only need to run it one time to initiate db.dataset table
```
Comment (caseyli92) [inline src/run_all.py:31]: It is giving this error:

Traceback (most recent call last):
  File "C:\Users\sli229\Projects\Digital-Twins\src\run_all.py", line 40, in <module>
    main(sample_polygon, module_to_log_level)
  File "C:\Users\sli229\Projects\Digital-Twins\src\run_all.py", line 24, in main
    module.main(selected_polygon_gdf, log_level=log_level)
  File "C:\Users\sli229\Projects\Digital-Twins\../NewZeaLiDAR\newzealidar\datasets.py", line 336, in main
    utils.map_dataset_name(engine, instructions_file)
  File "C:\Users\sli229\Projects\Digital-Twins\../NewZeaLiDAR\newzealidar\utils.py", line 231, in map_dataset_name
    with open(instructions_file, 'r') as f:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: 'configs\\instructions.json'

**Comment by LukeParky** [comment]
Hi Xander, could you also delete any of the parts within the DigitalTwin Lidar module that are unused? I believe most of it can be deleted but can you add this to the Pull Request?

File: src/flood_model/bg_flood_model.py (hunk 2 above)
Comment (xandercai) [reply src/flood_model/bg_flood_model.py]: done.

File: src/run_all.py (hunk 3 above)
Comment (xandercai) [reply src/run_all.py:28]: `.env` needs to be modified following the `.env.template`.

**Comment by xandercai** [comment]
The latest update (0920c13b9ef17fe1c231412d297dee4edce5fccb) fixed the issues that caused failures this afternoon.
1. GeoFabrics made a big update this month: the process is quite different from ours. So, I froze the GeoFabrics version to 0.10.23 to avoid significant changes until I get to understand the GeoFabrics update.
2. Since the NewZeaLiDAR package is installed by "pip: git+https" method, it may cause package-compatible issues. I clean the `environment.yml` to make sure the dependences of NewZeaLiDAR are compatible with Digital-Twins.

File: src/flood_model/bg_flood_model.py [hunk 4]
```diff
@@ -398,16 +398,13 @@ def run_bg_flood_model(
...
-    if resolution is None:
-        _, resolution = dem_metadata_in_db.get_hydro_dem_data_and_resolution(engine, catchment_area)
+    hydro_dem_path, _, _, resolution = get_dem_by_geometry(engine, catchment_area)
 
     # Prepare inputs for the BG-Flood Model
     prepare_bg_flood_model_inputs(
         bg_flood_dir=bg_flood_dir,
         model_output_path=model_output_path,
-        hydro_dem_path=hydro_dem_path,
+        hydro_dem_path=pathlib.Path(hydro_dem_path),
```
Comment (caseyli92) [inline src/flood_model/bg_flood_model.py:407]:
```suggestion
        hydro_dem_path=hydro_dem_path,
```

I believe it's beneficial to enhance the `get_dem_by_geometry()` function by incorporating clearly defined type hints for its output. This way, in scenarios like when the `hydro_dem_path` is utilized within the `prepare_bg_flood_model_inputs()` function, there will be no necessity to employ `pathlib.Path()`.

File: src/flood_model/bg_flood_model.py [hunk 5]
```diff
@@ -398,16 +398,13 @@ def run_bg_flood_model(
     # Get the valid BG-Flood Model directory
     bg_flood_dir = get_valid_bg_flood_dir()
     # Get the file path of the Hydro DEM for the catchment area
-    hydro_dem_path = dem_metadata_in_db.get_catchment_hydro_dem_filepath(engine, catchment_area)
-    # Use the resolution of the Hydro DEM if it is not provided
-    if resolution is None:
-        _, resolution = dem_metadata_in_db.get_hydro_dem_data_and_resolution(engine, catchment_area)
```
Comment (caseyli92) [inline src/flood_model/bg_flood_model.py:404]: As discussed in meeting, this resolution should be gathered from either the hydrological DEM or if you think it is a waste of resources then it should be read from a new column in the database. It shouldn't be deleted.

**Review by caseyli92 — COMMENTED** [review]
Hi @xandercai , Could you please eliminate any components within the DigitalTwin Lidar Module that are no longer relevant due to this code change?

**Comment by caseyli92** [comment]
Hi @xandercai @LukeParky, just wondering if the validation of existing hydro DEM for the AOI has been implemented yet.

File: src/flood_model/bg_flood_model.py [hunk 6]
```diff
@@ -398,16 +398,13 @@ def run_bg_flood_model(
...
+    hydro_dem_path, _, _, resolution = get_dem_by_geometry(engine, catchment_area)
```
Comment (LukeParky) [inline src/flood_model/bg_flood_model.py:401]:
```suggestion
    hydro_dem_path, _, _, resolution_of_dem = get_dem_by_geometry(engine, catchment_area)
    if resolution is None:
        resolution = resolution_of_dem
```
Idea of what @caseyli92 may be referring to, but disagreements on where the resolution should come from.
Resolution from either DEM or from user input for specifying output resolution or from database table

File: src/flood_model/bg_flood_model.py (hunk 5 above)
Comment (xandercai) [reply src/flood_model/bg_flood_model.py:404]: done

File: src/flood_model/bg_flood_model.py (hunk 6 above)
Comment (xandercai) [reply src/flood_model/bg_flood_model.py:401]: done

File: src/flood_model/bg_flood_model.py (hunk 4 above)
Comment (xandercai) [reply src/flood_model/bg_flood_model.py:407]: done

**Review by LukeParky — APPROVED** [review]
Looks good, ready to merge @xandercai

**Comment by LukeParky** [comment]
Ran tests locally. Approve merge

**Pull request merged**